evdev/uinput.py

```diff
@@ -281,7 +283,52 @@ def _verify(self):
             msg = 'uinput device name must not be longer than {} characters'
             raise UInputError(msg.format(_uinput.maxnamelen))
 
-    def _find_device(self):
+    def _find_device(self, fd):
+        '''
+        Tries to find the device node. Will delegate this task to one of
+        several platform-specific functions.
+        '''
+        try:
+            sysname = _uinput.get_sysname(fd)
+        except OSError:
+            # UI_GET_SYSNAME returned an error code. We're likely dealing with
+            # an old kernel. Guess the device based on the filesystem.
+            return self._find_device_fallback()
+
+        try:
+            if platform.system() == 'Linux':
+                return self._find_device_linux(sysname)
+        except OSError:
+            pass
+
+        # If we're not running or Linux or the above method fails for any reason,
+        # use the generic fallback method.
+        return self._find_device_fallback()
```

Comment:
sysname is only relevant if platform.system is linux, so it is probably better to wrap that in the condition as well.

I think this should be equivalent, haven't tested it yet:

```suggestion
    def _find_device(self, fd):
        '''
        Tries to find the device node. Will delegate this task to one of
        several platform-specific functions.
        '''
        if platform.system() == 'Linux':
            try:
                sysname = _uinput.get_sysname(fd)
                return self._find_device_linux(sysname)
            except OSError:
                # UI_GET_SYSNAME returned an error code. We're likely dealing with
                # an old kernel. Guess the device based on the filesystem.
                pass

        # If we're not running or Linux or the above method fails for any reason,
        # use the generic fallback method.
        return self._find_device_fallback()
```

what do you think?

Reply:
Sysname is relevant too if the platform is FreeBSD, it just gets a different meaning in that case. On Linux, it refers to some directory in `/sys/devices/virtual/input/`, but on FreeBSD it refers to some file in `/dev/input/`. I am not aware of other platforms that Linux and FreeBSD that support evdev.

The code I was originally going to propose was:

```py
try:
    sysname = _uinput.get_sysname(fd)
except OSError:
    # UI_GET_SYSNAME returned an error code. We're likely dealing with
    # an old kernel. Guess the device based on the filesystem.
    return self._find_device_fallback()

try:
    system = platform.system()
    if system == 'Linux':
        return self._find_device_linux(sysname)
    elif system = "FreeBSD":
        return self._find_device_freebsd(sysname)
except OSError:
    pass
```

... but then I realized I didn't have a FreeBSD VM running yet, and it would be bad form to propose some code based merely on "the documentation says it should work" without having actually tested it.

Now that I look at the above code again, it does seem like it could indeed be simplified; there are two try-catch blocks both of which have the same result of "skip the rest of the code and use `_find_device_linux()`". And unless code for handling the FreeBSD case is actually added, there is no reason not to move the `get_sysname()` within the check of whether we're running Linux.